### ISSUE TITLE
etcd-manager certificate rotation

### ISSUE THREAD

**Issue by rifelpet** [kubernetes/kops#8959]
When etcd-manager is first launched it issues certificates for each master node. These certs have a [hard-coded duration of one year](https://github.com/kopeio/etcd-manager/blob/1b7ffc96c132f6f1bed81f6ea67847cff6623d04/pkg/pki/certutils.go#L91). If the cert expires, apiserver will no longer be able to reach the local etcd members.

etcd-manager was first introduced in Kops 1.12 which was released May 15, 2019, just under a year ago. This means that many users will have an etcd outage soon unless we provide instructions on how to rotate the certs and remind them that this will happen. Long-term we can discuss etcd-manager providing automatic cert rotation, but in the short-term I think we need to at least document the manual steps required.

I've already seen outage reports in [#kops-users](https://kubernetes.slack.com/archives/C3QUFP0QM/p1587569013082800) and [twitter](https://twitter.com/kantrn/status/1253095873040379905) so it would be good to get a handle on this ASAP.

/priority critical-urgent

**Comment by marek-obuchowicz** [comment]
I can help with the docs. The process is:
 - delete certs using kops. This is currently broken, #8945 fixes this. Can we assume that documentation for etcd cert rotation can rely on this functionality, will it be released soon?
Without this fix, i'm deleting files directly from S3:
```
for cert in etcd-clients-ca etcd-manager-ca-events etcd-manager-ca-main etcd-peers-ca-events etcd-peers-ca-main; do
  aws s3 rm s3://${bucket}/${cluster}/pki/issued/${cert}/ --recursive
  aws s3 rm s3://korekontrol-kops-use1/p1.korekontrol.net/pki/private/${cert} --recursive
done
```

 - call kops to regenerate etcd certificates:
```
kops update cluster && kops update cluster --yes
```
 - get IPs of master nodes:
```
kubectl get nodes -owide | grep master
```
 - delete tls certs on each master node - connect via ssh and execute:
```
sudo find /mnt/vol-* -name server.* | xargs -I {} sudo rm {}
sudo find /mnt/vol-* -name me.* | xargs -I {} sudo rm {}
```
I think the step above can also be executed without SSH access, for example:
```
kubectl --namespace kube-system exec -ti etcd-manager-(...) bash -c "find /rootfs/mnt/vol-* (...)"
```

 - `kops rolling-update cluster --cloudonly --force --yes`
 - `kops export kubecfg`

**Comment by olemarkus** [comment]
> I can help with the docs. The process is:
> 
>     * delete certs using kops. This is currently broken, #8945 fixes this. Can we assume that documentation for etcd cert rotation can rely on this functionality, will it be released soon?
>       Without this fix, i'm deleting files directly from S3:
> 

I hope we can do a quick release to fix this. Should already be backported to 1.16.

> 
> ```
> kops update cluster && kops update cluster --yes
> ```
> 
>     * get IPs of master nodes:
> 
> ```
> kubectl get nodes -owide | grep master
> ```
> 
>     * delete tls certs on each master node - connect via ssh and execute:
> 
> 
> ```
> sudo find /mnt/vol-* -name server.* | xargs -I {} sudo rm {}
> sudo find /mnt/vol-* -name me.* | xargs -I {} sudo rm {}
> ```
> 
> I think the step above can also be executed without SSH access, for example:
> 
> ```
> kubectl --namespace kube-system exec -ti etcd-manager-(...) bash -c "find /rootfs/mnt/vol-* (...)"
> ```
>

It would be nice if nodeup could just replace these certs when there are newer in the secret store.
 
>     * Restart all cluster nodes:
> 
> 
> ```
> kops rolling-update cluster --cloudonly --force --yes
> ```
> 
>     * Not sure if this step is required (i found it in "rotating credentials file):
>       `kops export kubecfg`

You would only need to restart the masters and you could do that gracefully. etcd should work as long as the CA remains the same.

**Comment by sudermanjr** [comment]
> ```
> for cert in etcd-clients-ca etcd-manager-ca-events etcd-manager-ca-main etcd-peers-ca-events etcd-peers-ca-main; do
>   aws s3 rm s3://${bucket}/${cluster}/pki/issued/${cert}/ --recursive
>   aws s3 rm s3://${bucket}/${cluster}/pki/private/${cert}/ --recursive
> done
> ```

Taking a look at all the certs in those places in my clusters yields an expiration date that is 10 years out. I'm not sure how that could be, unless those are the wrong certs, or there is something else I'm missing here.

**Comment by rifelpet** [comment]
The certs and keys in S3 are only for the CA certificates which have a 10 year expiration. The individual certs with 1 year expirations are issued locally on each host, so nothing needs to be removed from S3 at the moment.

We just successfully rotated the etcd certs on our clusters and found these steps to work:

1. SSH into each master instance and delete the pki directory from each volume mount:
  ```
  sudo rm -rf /mnt/master-vol-*/pki   # mount paths may vary between cloud providers and distros. double check this path
  ```
2. Restart the two etcd-manager containers. Alternatively you can reboot the instance or terminate the instance in the autoscaling group
  ```
  sudo docker restart $(sudo docker ps -q -f "label=io.kubernetes.container.name=etcd-manager")
  ```

When the container restarts, etcd-manager will reissue the certs and rejoin the etcd cluster. You should be able to confirm membership with `etcdctl endpoint status` by following the instructions in the [docs](https://kops.sigs.k8s.io/operations/etcd_administration/#direct-data-access)

Doing that on one master at a time will avoid any apiserver downtime.

If anyone has had similar success or is willing to try these instructions please post here. I'm hoping we'll settle on a short term solution at kops office hours tomorrow.

**Comment by coderanger** [comment]
I put up a PR for etcd-manager to improve things. https://github.com/kopeio/etcd-manager/pull/309

**Comment by coderanger** [comment]
@rifelpet We used a similar approach, I did `mv pki pki-old` rather than `rm` but same idea. Everything came back cleanly. We had been on a prerelease of kops so I guess we got to experience this party a bit early.

**Comment by mitch000001** [comment]
I was just wondering how you would restore from a certificate rotation. Like, it is very easy to wipe all certs, but the more interesting and time consuming issue is the restoration of things afterwards. When we removed all (!) secrets and certs from a cluster and tried to start it over we had etcd complaining about old/bad certificates. We fixed it by restoring with `etcd-manager-ctl restore`, although I am not sure this is the correct way.

**Comment by olemarkus** [comment]
Just to stress what @rifelpet said: you do not need to do anything with the certs in the kops secret store (as in `kops get secret`). You only need to remove the certs from the etcd folders and then restart etcd.

**Comment by cbeneke** [comment]
Since this will affect people 1yr after they first installed kops in 1.12 a small hint how to check if (or better, when) you will be affected. SSH into any master and

```
sudo find /mnt/master-vol-* -name server.crt |xargs -I{} sudo openssl x509 -in {} -noout -text |grep 'Not After :'
```

This will print the end-times of the certs you have running. (The volume path might differ for different cloud-providers/distros.. double check :) ).

**Comment by mmerrill3** [comment]
Thanks @rifelpet, this process worked for me.  I just had an etcd outage in qa thankfully, applying this to prod.

**Comment by alasla** [comment]
I just had the same thing happen.  Everything ground to a halt because some of the keys/certs on the masters hit their one year timer.  Dumping the pki directories as @rifelpet suggested seems to have done the trick.  Thanks @rifelpet .

**Comment by cmanzi** [comment]
Looks like etcd-manager will now eagerly renew certs that are expiring as of the release last week: https://github.com/kopeio/etcd-manager/releases. Has anyone tested this out with Kops?

Also, for those of us who are not using 10-year CAs and plan on rotating them, has anyone devised a process for doing so non-disruptively? From what I've seen, rotating the CA and issuing new certs creates a problem where nodes do not trust each others' CAs, effectively breaking the etcd cluster until all nodes are rotated.

**Comment by ari-becker** [comment]
> Also, for those of us who are not using 10-year CAs and plan on rotating them, has anyone devised a process for doing so non-disruptively? From what I've seen, rotating the CA and issuing new certs creates a problem where nodes do not trust each others' CAs, effectively breaking the etcd cluster until all nodes are rotated.

Short answer: no.

Longer answer: at the end of the day, the master servers are ordinary Linux servers running the various Kubernetes components, and the Kops state store is not encrypted. Nothing prevents the user from operating outside the `kops` command line tool.

But I really don't understand how non-disruptive credential rotation is not a supported workflow. It makes it impossible to recommend Kops for production clusters - the ability to rotate credentials is a critical part of any security posture.

**Comment by olemarkus** [comment]
> > Also, for those of us who are not using 10-year CAs and plan on rotating them, has anyone devised a process for doing so non-disruptively? From what I've seen, rotating the CA and issuing new certs creates a problem where nodes do not trust each others' CAs, effectively breaking the etcd cluster until all nodes are rotated.
> 
> Short answer: no.
> 
> Longer answer: at the end of the day, the master servers are ordinary Linux servers running the various Kubernetes components, and the Kops state store is not encrypted. Nothing prevents the user from operating outside the `kops` command line tool.
> 
> But I really don't understand how non-disruptive credential rotation is not a supported workflow. It makes it impossible to recommend Kops for production clusters - the ability to rotate credentials is a critical part of any security posture.

If you have an idea on how to mom-disruptively rotate a CA, I'd be happy to take a look at that.

Kops biggest sin from my side in this regard is that the CA private key is in the state store. That is something I hope to do something about at some point.

**Comment by cmanzi** [comment]
> If you have an idea on how to mom-disruptively rotate a CA, I'd be happy to take a look at that.
I have yet to test this (though I will soon), but theoretically one should be able to:
1. Generate a new CA (we sign our CAs with a common root CA)
2. Use current CA and new CA to cross-sign a new set of Etcd client/server keypairs
3. Replace current cert pairs with cross-signed pairs
4. Deprecate old CA and continue rotation using new CA

This of course depends on Etcd being able to then validate the cross-signed certs.

> Kops biggest sin from my side in this regard is that the CA private key is in the state store. That is something I hope to do something about at some point.
Yes, this can be mitigated by the use of KMS encryption and restrictive bucket policies, but is still worrying.

It would be quite nice if we could get a native Vault integration. That way you could just request intermediate CAs/certs directly from Vault.

**Comment by marek-obuchowicz** [comment]
@justinsb I see in etcd-manager's changelog that this issue has been adressed. Do you happen to know anything about that? Is upgrading `etcd-manager` just the only step needed? What is the correct workflow to update it (ie. `kops edit cluster && kops update cluster --yes`, rotate masters with/without `--cloudonly`, anything else)?

I have the feeling that this issue might be found by many kops users "too late", so after etcd certs expire and clusters are down. A clear, step-by-step instruction could help achieving shorter recovery times.

**Comment by rifelpet** [comment]
Instructions and more information is available in the [advisory here](https://kops.sigs.k8s.io/advisories/etcd-manager-certificate-expiration/).

**Comment by GMartinez-Sisti** [comment]
Upgrading from 1.16.0 to 1.16.2 doesn't work for me. The new master is not able to join the etcd events cluster. The other masters complain with:

```2020-05-07 15:48:51.792598 I | embed: rejected connection from "10.200.20.115:23166" (error "tls: \"10.200.20.115\" does not match any of DNSNames [\"etcd-events-b.internal.cluster.k8s.local\"]", ServerName "etcd-events-a.internal.cluster.k8s.local", IPAddresses ["127.0.0.1"], DNSNames ["etcd-events-b.internal.cluster.k8s.local"])```

etcd main updated fine, but events is not.

After rolling back, the same issue is happening, so now I'm 1 master down until I can figure this out.

I'm running:
- Kops 1.16.0 or 1.16.2
- Kubernetes 1.16.7 in AWS
- etcd  3.3.10 with enableEtcdTLS set to true
- etcd-manager 3.0.20191025

I don't want to hijack this thread, but I'm guessing a lot of people will upgrade kops to prevent the certificate issue and might run into the same issue.

**Comment by olemarkus** [comment]
The fix for certificate rotation is that etcd-manager is upgraded to version 3.0.20200428. Since you have set a custom etcd-manager version, you need to fix this yourself or use kops default.

**Comment by GMartinez-Sisti** [comment]
I had to upgrade to use [Calico v3](https://github.com/kubernetes/kops/blob/master/docs/calico-v3.md). It's a supported procedure by kops, so I'm sharing for awareness and to see if someone had the same problem.

**Comment by flmmartins** [comment]
Hello everyone,
I just did the upgrade to 1.16.2 & etcd-manager:3.0.20200429. Everything went okey.
Although I am watching Kops Github I didn't detect this.

Is there a way we can be informed about critical changes or releases by Kops? Or a specific place one should follow? I guess this would help a lot in case we have critical situations.

**Comment by olemarkus** [comment]
It was mentined here: https://github.com/kubernetes/kops/releases/tag/v1.16.2
If you follow github releases, you'd get this. I admit that the seriousness of this issue could be emphasised more on that page though.
Id create a separate issue on pushing info like this.

I actually think this particular issue can be closed now.

**Comment by atgaurav** [comment]
I am validating the expiry of the certs but it seems there are multiple certs.

The command mentioned by @cbeneke  lists certs in etcd-manager-events-etcd.
Then there are certs in directory /etc/kubernetes/pki/

Which one causes the specified disruption here, can anyone shed some more light here please?

**Comment by olemarkus** [comment]
The certs in `/etc/kubernetes/pki/` are outside the scope of this issue. These certs are certs managed by kops, not etcd-manager.
The certs in this issue are only the ones in the etcd-manager mounts.

**Comment by atgaurav** [comment]
@olemarkus Thanks for confirmation, just rephrasing here
Certs in the directory "/etc/kubernetes/pki/" are managed by Kops so they are out of scope of this issue.

But certs in the directory are the impacted ones:
```
/mnt/master-vol-******/pki/etcd-cluster-token-etcd/clients/server.crt
/mnt/master-vol-******/pki/etcd-cluster-token-etcd-events/clients/server.crt
/mnt/master-vol-******/pki/etcd-cluster-token-etcd/peers/me.crt
/mnt/master-vol-******/pki/etcd-cluster-token-etcd-events/peers/me.crt
```

Even the CA certs in the above mentioned directory are out of scope.

**Comment by duncanhall** [comment]
**_As a warning to others (and a request for any further help:_**

My master node started failing to communicate with worker nodes precisely 1 year after it was created. I arrived at this thread, and followed the advisory notes to confirm that I was affected by this issue. I `mv`'d the `/mnt/master-*/pki` directories and restarted my etcd-manager containers, but this seems to have simply made things worse. 

`kubelet` logs now show continuous failures around SSL connections:

```
Failed to list *v1.Pod: Get https://127.0.0.1/api/v1/pods?fieldSelector=spec.nodeName%3Dip-XYZ.compute.internal&limit=500&resourceVersion=0: dial tcp 127.0.0.1:443: connect: connection refused
Failed to list *v1.Node: Get https://127.0.0.1/api/v1/nodes?fieldSelector=metadata.name%3Dip-XYZ.compute.internal&limit=500&resourceVersion=0: dial tcp 127.0.0.1:443: connect: connection refused
```

Looking through `etcd(-manager)` logs it seems that etcd can no longer start because of certificate issues:

```
I0813 13:27:43.695789    2136 controller.go:173] starting controller iteration
W0813 13:27:43.695871    2136 controller.go:149] unexpected error running etcd cluster reconciliation loop: cannot find self "etcd-a" in list of peers []
I0813 13:27:46.035303    2136 peers.go:281] connecting to peer "etcd-a" with TLS policy, servername="etcd-manager-server-etcd-a"
W0813 13:27:46.038572    2136 peers.go:325] unable to grpc-ping discovered peer 172.20.60.88:3996: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid"
I0813 13:27:46.038611    2136 peers.go:347] was not able to connect to peer etcd-a: map[172.20.60.88:3996:true]
```

So now my entire cluster is unable to communicate and `kubectl` is unable to function:

```
$ kubectl cluster-info
The connection to the server localhost:8080 was refused - did you specify the right host or port?
```

**Comment by olemarkus** [comment]
Is that from running `kubectl` locally? It looks a lot like the error you'd expect if you didn't have a local kube config.

**Comment by duncanhall** [comment]
> Is that from running `kubectl` locally? It looks a lot like the error you'd expect if you didn't have a local kube config.

Thanks @olemarkus - it's the same response from local and from running on the master. This is what I would expect given that api-server is unable to start because etcd is unable to start.

**Comment by hakman** [comment]
I would suggest to try to upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531` by editing the manifest files and restarting the etcd pods:
```
/etc/kubernetes/manifests/etcd-events.manifest
/etc/kubernetes/manifests/etcd-main.manifest
```

**Comment by duncanhall** [comment]
> I would suggest to try to upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531` by editing the manifest files and restarting the etcd pods:
> 
> ```
> /etc/kubernetes/manifests/etcd-events.manifest
> /etc/kubernetes/manifests/etcd-main.manifest
> ```

Thanks @hakman this was a very helpful nudge!  etcd is now back up and running which is a great start, but wiping and regenerating the `/pki` dir still seems to be causing issues:

Although etcd is back up, apiserver is still failure cycling and cannot start. The last log before it dies and tries to restart shows:

```
Unable to create storage backend: config (&{etcd3 /registry [https://127.0.0.1:4002] /etc/kubernetes/pki/kube-apiserver/etcd-client.key /etc/kubernetes/pki/kube-apiserver/etcd-client.crt /etc/kubernetes/pki/kube-apiserver/etcd-ca.crt true true 1000 {0xc420143560 0xc4201435f0} <nil> 5m0s 1m0s}), err (dial tcp 127.0.0.1:4002: connect: connection refused)
```

From the error above it feels as though simply wiping/regenerating the `/mnt/master-vol-*/pki` directory is only half the story?

**Comment by hakman** [comment]
Are you sure that both main and events etcd clusters are starting ok?

**Comment by duncanhall** [comment]
Thanks @hakman - correct again, it seems etcd-events can't start because the `data-events` directory does not exist. Is this likely a consequence of updating the etcd-manager image? Is there a solution to this I'm missing?

```
error initializing etcd server: error reading source directory /rootfs/mnt/master-vol-09da1a418d361ef92/var/etcd/data-events: open /rootfs/mnt/master-vol-09da1a418d361ef92/var/etcd/data-events: no such file or directory
```

**Comment by hakman** [comment]
`/rootfs/mnt/master-vol-09da1a418d361ef92` is an EBS disk attached to the instance. Either it's not mounted anymore, or the data is gone (which is not a good thing).

**Comment by duncanhall** [comment]
> `/rootfs/mnt/master-vol-09da1a418d361ef92` is an EBS disk attached to the instance. Either it's not mounted anymore, or the data is gone (which is not a good thing).

Thanks @hakman - the EBS volume is still mounted so it does seem that the specific `var/etcd/data-events` directory is gone. 

I am not 100% clear on the relevance of the contents of this directory. Given that my master is currently unable to communicate with any worker nodes anyway, I'm willing to take some risk / data loss to get the cluster back to health.

 - Would simply creating an empty directory at `/mnt/master-vol-xyz/var/etcd/data-events` be enough to get etcd-events running without destructive consequences?
 - I also have an option to restore etcd-events from S3 backup - although they are quite stale. How destructive could it be to try to restore etcd-events from a pretty old backup?

**Comment by hakman** [comment]
At the moment, I don't know of any other way to restore the date than to use the backup. So... your choice if you want to try it or not.

**Comment by duncanhall** [comment]
For anyone hitting the same issue I did above, simply creating an empty dir at `/mnt/master-vol-xyz/var/etcd/data-events` was enough to get etcd-events to start correctly.

I did lose some historical data through using a stale backup, but more importantly my cluster is back to health. Thank you for your help @hakman !

**Comment by hakman** [comment]
Glad it all worked out @duncanhall :)

**Comment by kylerisse** [comment]
> The certs and keys in S3 are only for the CA certificates which have a 10 year expiration. The individual certs with 1 year expirations are issued locally on each host, so nothing needs to be removed from S3 at the moment.
> 
> We just successfully rotated the etcd certs on our clusters and found these steps to work:
> 
> 1. SSH into each master instance and delete the pki directory from each volume mount:
> 
> ```
> sudo rm -rf /mnt/master-vol-*/pki   # mount paths may vary between cloud providers and distros. double check this path
> ```
> 
> 1. Restart the two etcd-manager containers. Alternatively you can reboot the instance or terminate the instance in the autoscaling group
> 
> ```
> sudo docker restart $(sudo docker ps -q -f "label=io.kubernetes.container.name=etcd-manager")
> ```
> 
> When the container restarts, etcd-manager will reissue the certs and rejoin the etcd cluster. You should be able to confirm membership with `etcdctl endpoint status` by following the instructions in the [docs](https://kops.sigs.k8s.io/operations/etcd_administration/#direct-data-access)
> 
> Doing that on one master at a time will avoid any apiserver downtime.
> 
> If anyone has had similar success or is willing to try these instructions please post here. I'm hoping we'll settle on a short term solution at kops office hours tomorrow.

I can confirm that this procedure did resolve the etcd issues on 3 different clusters. Thanks a lot!

**Comment by accnops** [comment]
I had this issue today, this issue and [this page](https://github.com/kubernetes/kops/blob/master/docs/rotate-secrets.md) helped me fix it. I'm commenting here to make this page more visible in search engine results.

The symptom: Kubernetes API was suddenly completely down.
Luckily, I had SSH access to the master node. There, I checked some logs, using `sudo docker ps` and then `sudo docker logs [container-id]`.

I found following errors:

On k8s_etcd-manager_etcd-manager-main-ip-... 
```
I0914 12:23:15.143808   11520 controller.go:173] starting controller iteration
W0914 12:23:15.143855   11520 controller.go:149] unexpected error running etcd cluster reconciliation loop: cannot find self "etcd-b" in list of peers []
I0914 12:23:19.598312   11520 peers.go:281] connecting to peer "etcd-b" with TLS policy, servername="etcd-manager-server-etcd-b"
W0914 12:23:19.603640   11520 peers.go:325] unable to grpc-ping discovered peer 172.20.35.57:3996: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
I0914 12:23:19.603783   11520 peers.go:347] was not able to connect to peer etcd-b: map[172.20.35.57:3996:true]
W0914 12:23:19.603894   11520 peers.go:215] unexpected error from peer intercommunications: unable to connect to peer etcd-b
I0914 12:23:24.605336   11520 peers.go:281] connecting to peer "etcd-b" with TLS policy, servername="etcd-manager-server-etcd-b"
W0914 12:23:24.608827   11520 peers.go:325] unable to grpc-ping discovered peer 172.20.35.57:3996: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
I0914 12:23:24.608865   11520 peers.go:347] was not able to connect to peer etcd-b: map[172.20.35.57:3996:true]
W0914 12:23:24.608892   11520 peers.go:215] unexpected error from peer intercommunications: unable to connect to peer etcd-b
```

This log was repeated over and over.

The API server container was crashing with error:
```
F0913 19:33:20.945576       1 storage_decorator.go:57] Unable to create storage backend: config (&{etcd3 /registry [https://127.0.0.1:4001] /etc/kubernetes/pki/kube-apiserver/etcd-client.key /etc/kubernetes/pki/kube-apiserver/etcd-client.crt /etc/kubernetes/pki/kube-apiserver/etcd-ca.crt true true 1000 0xc420170ea0 <nil> 5m0s 1m0s}), err (context deadline exceeded)
```
Suggesting there was something wrong with etcd.

**Comment by jeromeroserio** [comment]
I had similar issue couple of days ago where I can't connect to my Kubernetes cluster using `kubectl`. It was giving  The connection to the server api.xxxxx.xxxxxxxx.com was refused - did you specify the right host or port?"

I have a 4 node Kubernetes cluster with 1 x master and 3 x worker nodes hosted on AWS. This Kubernetes cluster was setup using kops by a colleague of mine who no longer available for comment.

I connected to master server and `kubectl` wasn't working there as well. I found that kube-apiserver docker container was restarting and one of the error was 

`W0914 16:29:25.761524 1 clientconn.go:1251] grpc: addrConn.createTransport failed to connect to {127.0.0.1:4001 0 }. Err :connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid". Reconnecting... F0914 16:29:29.319785 1 storage_decorator.go:57] Unable to create storage backend: config (&{etcd3 /registry {[https://127.0.0.1:4001] /etc/kubernetes/pki/kube-apiserver/etcd-client.key /etc/kubernetes/pki/kube-apiserver/etcd-client.crt /etc/kubernetes/pki/kube-apiserver/etcd-ca.crt} false true 0xc000266d80 apiextensions.k8s.io/v1beta1 5m0s 1m0s}), err (context deadline exceeded)`

At this time, I was unaware of ETCD certificates and thought that only kube-apiserver certificate has expired. 
I found the kube-apiserver certificate  `/etc/kubernetes/pki/kube-apiserver/etcd-client.crt` which expired in July 2020. So I manually generated the certificate using openssl and then restarted kube-apiserver docker. Now I received different error

`F0916 08:09:56.753538 1 storage_decorator.go:57] Unable to create storage backend: config (&{etcd3 /registry {[https:--127.0.0.1:4001] /etc/kubernetes/pki/kube-apiserver/etcd-client.key /etc/kubernetes/pki/kube-apiserver/etcd-client.crt /etc/kubernetes/pki/kube-apiserver/etcd-ca.crt} false true 0xc00095f050 apiextensions.k8s.io/v1beta1 5m0s 1m0s}), err (tls: private key does not match public key)`

**Then few mins ago (16th Sep), I realized that I only replaced the certificate but not the key. So I replaced the key and then restarted kube-apiserver docker. Now kube-apiserver is back online and I am able to connect using `kubectl`.**

In the meantime, I asked this question online yesterday (15th Sep)  and received response to check on ETCD certificates. Then I found that ETCD certificates had expired as well. 

**So I followed the workaround steps from https://kops.sigs.k8s.io/advisories/etcd-manager-certificate-expiration/ and renewed the certificates.**

Now my etcd is not working and giving below errors

Note: xxx-xx-xx-xxx is the internal IP address of the master server.

root@ip-xxx-xx-xx-xxx:~# docker logs <etcd-manager-main container> --tail 20
`I0916 14:41:41.205763    8221 controller.go:173] starting controller iteration
W0916 14:41:41.205801    8221 controller.go:149] unexpected error running etcd cluster reconciliation loop: cannot find self "etcd-a" in list of peers []
I0916 14:41:45.352008    8221 peers.go:281] connecting to peer "etcd-a" with TLS policy, servername="etcd-manager-server-etcd-a"
W0916 14:41:45.354445    8221 peers.go:325] unable to grpc-ping discovered peer xxx.xx.xx.xxx:3996: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
I0916 14:41:45.354461    8221 peers.go:347] was not able to connect to peer etcd-a: map[xxx.xx.xx.xxx:3996:true]
W0916 14:41:45.354473    8221 peers.go:215] unexpected error from peer intercommunications: unable to connect to peer etcd-a`

root@ip-xxx-xx-xx-xxx:~# docker logs <etcd-manager-events container> --tail 20
`I0916 14:42:41.106654    8316 controller.go:173] starting controller iteration
W0916 14:42:41.106692    8316 controller.go:149] unexpected error running etcd cluster reconciliation loop: cannot find self "etcd-events-a" in list of peers []
I0916 14:42:45.294682    8316 peers.go:281] connecting to peer "etcd-events-a" with TLS policy, servername="etcd-manager-server-etcd-events-a"
W0916 14:42:45.297094    8316 peers.go:325] unable to grpc-ping discovered peer xxx.xx.xx.xxx:3997: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
I0916 14:42:45.297117    8316 peers.go:347] was not able to connect to peer etcd-events-a: map[xxx.xx.xx.xxx:3997:true]
W0916 14:42:45.297130    8316 peers.go:215] unexpected error from peer intercommunications: unable to connect to peer etcd-events-a`

I checked the ETCD certificates and the alternate names (DNS) entries are like `etcd-a.internal.abc.xyz.com` and `etcd-events-a.internal.abc.xyz.com`. Similar entries are there in /etc/hosts file. I am not sure if the etcd is looking for `etcd-a` and `etcd-events-a` instead of FQDN?

My etcd-manager image is `kopeio/etcd-manager:3.0.20190516` 
Kubernetes version: v1.14.1
Docker version: 18.6.3

Any help to fix the issue with etcd would be really appreciated.

Many thanks.

**Comment by accnops** [comment]
> I had similar issue couple of days ago where I can't connect to my Kubernetes cluster using `kubectl`. It was giving The connection to the server api.xxxxx.xxxxxxxx.com was refused - did you specify the right host or port?"
> 
> I have a 4 node Kubernetes cluster with 1 x master and 3 x worker nodes hosted on AWS. This Kubernetes cluster was setup using kops by a colleague of mine who no longer available for comment.
> 
> I connected to master server and `kubectl` wasn't working there as well. I found that kube-apiserver docker container was restarting and one of the error was
> 
> `W0914 16:29:25.761524 1 clientconn.go:1251] grpc: addrConn.createTransport failed to connect to {127.0.0.1:4001 0 }. Err :connection error: desc = "transport: authentication handshake failed: x509: certificate has expired or is not yet valid". Reconnecting... F0914 16:29:29.319785 1 storage_decorator.go:57] Unable to create storage backend: config (&{etcd3 /registry {[https://127.0.0.1:4001] /etc/kubernetes/pki/kube-apiserver/etcd-client.key /etc/kubernetes/pki/kube-apiserver/etcd-client.crt /etc/kubernetes/pki/kube-apiserver/etcd-ca.crt} false true 0xc000266d80 apiextensions.k8s.io/v1beta1 5m0s 1m0s}), err (context deadline exceeded)`
> 
> At this time, I was unaware of ETCD certificates and thought that only kube-apiserver certificate has expired.
> I found the kube-apiserver certificate `/etc/kubernetes/pki/kube-apiserver/etcd-client.crt` which expired in July 2020. So I manually generated the certificate using openssl and then restarted kube-apiserver docker. Now I received different error
> 
> `F0916 08:09:56.753538 1 storage_decorator.go:57] Unable to create storage backend: config (&{etcd3 /registry {[https:--127.0.0.1:4001] /etc/kubernetes/pki/kube-apiserver/etcd-client.key /etc/kubernetes/pki/kube-apiserver/etcd-client.crt /etc/kubernetes/pki/kube-apiserver/etcd-ca.crt} false true 0xc00095f050 apiextensions.k8s.io/v1beta1 5m0s 1m0s}), err (tls: private key does not match public key)`
> 
> **Then few mins ago (16th Sep), I realized that I only replaced the certificate but not the key. So I replaced the key and then restarted kube-apiserver docker. Now kube-apiserver is back online and I am able to connect using `kubectl`.**
> 
> In the meantime, I asked this question online yesterday (15th Sep) and received response to check on ETCD certificates. Then I found that ETCD certificates had expired as well.
> 
> **So I followed the workaround steps from https://kops.sigs.k8s.io/advisories/etcd-manager-certificate-expiration/ and renewed the certificates.**
> 
> Now my etcd is not working and giving below errors
> 
> Note: xxx-xx-xx-xxx is the internal IP address of the master server.
> 
> root@ip-xxx-xx-xx-xxx:~# docker logs --tail 20
> `I0916 14:41:41.205763 8221 controller.go:173] starting controller iteration W0916 14:41:41.205801 8221 controller.go:149] unexpected error running etcd cluster reconciliation loop: cannot find self "etcd-a" in list of peers [] I0916 14:41:45.352008 8221 peers.go:281] connecting to peer "etcd-a" with TLS policy, servername="etcd-manager-server-etcd-a" W0916 14:41:45.354445 8221 peers.go:325] unable to grpc-ping discovered peer xxx.xx.xx.xxx:3996: rpc error: code = Unavailable desc = all SubConns are in TransientFailure I0916 14:41:45.354461 8221 peers.go:347] was not able to connect to peer etcd-a: map[xxx.xx.xx.xxx:3996:true] W0916 14:41:45.354473 8221 peers.go:215] unexpected error from peer intercommunications: unable to connect to peer etcd-a`
> 
> root@ip-xxx-xx-xx-xxx:~# docker logs --tail 20
> `I0916 14:42:41.106654 8316 controller.go:173] starting controller iteration W0916 14:42:41.106692 8316 controller.go:149] unexpected error running etcd cluster reconciliation loop: cannot find self "etcd-events-a" in list of peers [] I0916 14:42:45.294682 8316 peers.go:281] connecting to peer "etcd-events-a" with TLS policy, servername="etcd-manager-server-etcd-events-a" W0916 14:42:45.297094 8316 peers.go:325] unable to grpc-ping discovered peer xxx.xx.xx.xxx:3997: rpc error: code = Unavailable desc = all SubConns are in TransientFailure I0916 14:42:45.297117 8316 peers.go:347] was not able to connect to peer etcd-events-a: map[xxx.xx.xx.xxx:3997:true] W0916 14:42:45.297130 8316 peers.go:215] unexpected error from peer intercommunications: unable to connect to peer etcd-events-a`
> 
> I checked the ETCD certificates and the alternate names (DNS) entries are like `etcd-a.internal.abc.xyz.com` and `etcd-events-a.internal.abc.xyz.com`. Similar entries are there in /etc/hosts file. I am not sure if the etcd is looking for `etcd-a` and `etcd-events-a` instead of FQDN?
> 
> My etcd-manager image is `kopeio/etcd-manager:3.0.20190516`
> Kubernetes version: v1.14.1
> Docker version: 18.6.3
> 
> Any help to fix the issue with etcd would be really appreciated.
> 
> Many thanks.

I had the same issue. Following [these instructions](https://github.com/kubernetes/kops/blob/master/docs/rotate-secrets.md) helped me a lot!

**Comment by jeromeroserio** [comment]
Thanks @accnops. 
I went through the instructions at https://github.com/kubernetes/kops/blob/master/docs/rotate-secrets.md and as  it says at the beginning it is a disruptive procedure. 

Currently my prod services are running OK. I understand if any POD has issues, then I won't be able to perform any management  tasks as ETCD is not working. So is there any way to bring ECTD back online without disruptive process?

Many Thanks.

**Comment by accnops** [comment]
> Thanks @accnops.
> I went through the instructions at https://github.com/kubernetes/kops/blob/master/docs/rotate-secrets.md and as it says at the beginning it is a disruptive procedure.
> 
> Currently my prod services are running OK. I understand if any POD has issues, then I won't be able to perform any management tasks as ETCD is not working. So is there any way to bring ECTD back online without disruptive process?
> 
> Many Thanks.

Yes, it will recreate a master and worker nodes. It's like a cluster version upgrade.

**Comment by jeromeroserio** [comment]
> I would suggest to try to upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531` by editing the manifest files and restarting the etcd pods:
> 
> ```
> /etc/kubernetes/manifests/etcd-events.manifest
> /etc/kubernetes/manifests/etcd-main.manifest
> ```

@hakman : I also experience similar issue like @duncanhall where I couldn't communicate using kubectl because kupe-apiserver certificate expired. So I manually (using openssl) created the certificate and replaced `/etc/kubernetes/pki/kube-apiserver/etcd-client.crt` and `/etc/kubernetes/pki/kube-apiserver/etcd-client.key` after that kube-apiserver docker started working and I am able to connect using kubectl. 

Then I followed the workaround from [https://kops.sigs.k8s.io/advisories/etcd-manager-certificate-expiration/](url) to generate certs for etcd which are stored at /mnt/master-vol-*/pki. Now I have renewed certs at /mnt/master-vol-*/pki but both etcd-manager-main and etcd-manager-events are giving error saying `not able to connect to peer etcd-a` (for main) and `etcd-events-a` (for events). More details on errors are posted on my comments on this thread two days ago.

@hakman: I have not upgraded etcd-manager to `kopeio/etcd-manager:3.0.20200531` yet. I am currently on `kopeio/etcd-manager:3.0.20190516`. As of now, I don't see /var folder under /mnt/master-vol-*/. I have only `data`, `pki`, `state` and `lost+found` folders. 

Is it ok for me to upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531`?

@duncanhall : You mentioned that you restored data from a stale backup, could you please provide what folders data you had to restore? 

Thanks @hakman @duncanhall

**Comment by jeromeroserio** [comment]
> > Thanks @accnops.
> > I went through the instructions at https://github.com/kubernetes/kops/blob/master/docs/rotate-secrets.md and as it says at the beginning it is a disruptive procedure.
> > Currently my prod services are running OK. I understand if any POD has issues, then I won't be able to perform any management tasks as ETCD is not working. So is there any way to bring ECTD back online without disruptive process?
> > Many Thanks.
> 
> Yes, it will recreate a master and worker nodes. It's like a cluster version upgrade.

Thanks @accnops for your response. I am trying to avoid the re-creation of master and worker nodes if possible.

**Comment by jeromeroserio** [comment]
> > I would suggest to try to upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531` by editing the manifest files and restarting the etcd pods:
> > ```
> > /etc/kubernetes/manifests/etcd-events.manifest
> > /etc/kubernetes/manifests/etcd-main.manifest
> > ```
> 
> @hakman : I also experience similar issue like @duncanhall where I couldn't communicate using kubectl because kupe-apiserver certificate expired. So I manually (using openssl) created the certificate and replaced `/etc/kubernetes/pki/kube-apiserver/etcd-client.crt` and `/etc/kubernetes/pki/kube-apiserver/etcd-client.key` after that kube-apiserver docker started working and I am able to connect using kubectl.
> 
> Then I followed the workaround from [https://kops.sigs.k8s.io/advisories/etcd-manager-certificate-expiration/](url) to generate certs for etcd which are stored at /mnt/master-vol-_/pki. Now I have renewed certs at /mnt/master-vol-_/pki but both etcd-manager-main and etcd-manager-events are giving error saying `not able to connect to peer etcd-a` (for main) and `etcd-events-a` (for events). More details on errors are posted on my comments on this thread two days ago.
> 
> @hakman: I have not upgraded etcd-manager to `kopeio/etcd-manager:3.0.20200531` yet. I am currently on `kopeio/etcd-manager:3.0.20190516`. As of now, I don't see /var folder under /mnt/master-vol-*/. I have only `data`, `pki`, `state` and `lost+found` folders.
> 
> Is it ok for me to upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531`?
> 
> @duncanhall : You mentioned that you restored data from a stale backup, could you please provide what folders data you had to restore?
> 
> Thanks @hakman @duncanhall

**Update to my previous comment**

I have two other Kubernertes clusters setup using kops in dev/qa. These clusters are running Kubernetes version v1.11.1 and do not have etcd-manager. I checked the path on the master volumes and found the path to be 

for `etcd.manifest` --> `/mnt/master-vol-***/var/etcd/data/member/` which has folders `snap` and `wal`
for `etcd-events.manifest` --> `/mnt/master-vol-****/var/etcd/data-events/member/` which has folders `snap` and `wal`

So in my production cluster which has etcd-manager installed (`kopeio/etcd-manager:3.0.20190516`), folders under /mnt/master-vol/*** are  `data`, `pki`, `state` and `lost+found` folders. 

`/mnt/master-vol-****/data/<token>/member/` which has folders `snap` and `wal`

Just want to make sure below steps are correct to fix the issue?

- Backup existing `data` and `data-events` folders (or all the contents under /mnt/master-vol-*)
- Upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531` and check if it starts correctly and resolves the issue.
- If it complains about missing the folder ``/mnt/master-vol-***/var/etcd/data` or `/mnt/master-vol-***/var/etcd/data-events`. then copy the contents from the backup.

Appreciate if anyone could please comment on the above or if I am missing any steps.

Many thanks.

**Comment by jeromeroserio** [comment]
> > > I would suggest to try to upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531` by editing the manifest files and restarting the etcd pods:
> > > ```
> > > /etc/kubernetes/manifests/etcd-events.manifest
> > > /etc/kubernetes/manifests/etcd-main.manifest
> > > ```
> > 
> > 
> > @hakman : I also experience similar issue like @duncanhall where I couldn't communicate using kubectl because kupe-apiserver certificate expired. So I manually (using openssl) created the certificate and replaced `/etc/kubernetes/pki/kube-apiserver/etcd-client.crt` and `/etc/kubernetes/pki/kube-apiserver/etcd-client.key` after that kube-apiserver docker started working and I am able to connect using kubectl.
> > Then I followed the workaround from [https://kops.sigs.k8s.io/advisories/etcd-manager-certificate-expiration/](url) to generate certs for etcd which are stored at /mnt/master-vol-_/pki. Now I have renewed certs at /mnt/master-vol-_/pki but both etcd-manager-main and etcd-manager-events are giving error saying `not able to connect to peer etcd-a` (for main) and `etcd-events-a` (for events). More details on errors are posted on my comments on this thread two days ago.
> > @hakman: I have not upgraded etcd-manager to `kopeio/etcd-manager:3.0.20200531` yet. I am currently on `kopeio/etcd-manager:3.0.20190516`. As of now, I don't see /var folder under /mnt/master-vol-*/. I have only `data`, `pki`, `state` and `lost+found` folders.
> > Is it ok for me to upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531`?
> > @duncanhall : You mentioned that you restored data from a stale backup, could you please provide what folders data you had to restore?
> > Thanks @hakman @duncanhall
> 
> **Update to my previous comment**
> 
> I have two other Kubernertes clusters setup using kops in dev/qa. These clusters are running Kubernetes version v1.11.1 and do not have etcd-manager. I checked the path on the master volumes and found the path to be
> 
> for `etcd.manifest` --> `/mnt/master-vol-***/var/etcd/data/member/` which has folders `snap` and `wal`
> for `etcd-events.manifest` --> `/mnt/master-vol-****/var/etcd/data-events/member/` which has folders `snap` and `wal`
> 
> So in my production cluster which has etcd-manager installed (`kopeio/etcd-manager:3.0.20190516`), folders under /mnt/master-vol/*** are `data`, `pki`, `state` and `lost+found` folders.
> 
> `/mnt/master-vol-****/data/<token>/member/` which has folders `snap` and `wal`
> 
> Just want to make sure below steps are correct to fix the issue?
> 
> * Backup existing `data` and `data-events` folders (or all the contents under /mnt/master-vol-*)
> * Upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531` and check if it starts correctly and resolves the issue.
> * If it complains about missing the folder ``/mnt/master-vol-_**/var/etcd/data`or`/mnt/master-vol-**_/var/etcd/data-events`. then copy the contents from the backup.
> 
> Appreciate if anyone could please comment on the above or if I am missing any steps.
> 
> Many thanks.

**Update:**

I have upgraded etcd-manager to `kopeio/etcd-manager:3.0.20200531` which resolved the issue. Cluster is back to healthy state. Many thanks to @hakman.

**Comment by lboix** [comment]
> The certs and keys in S3 are only for the CA certificates which have a 10 year expiration. The individual certs with 1 year expirations are issued locally on each host, so nothing needs to be removed from S3 at the moment.
> 
> We just successfully rotated the etcd certs on our clusters and found these steps to work:
> 
> 1. SSH into each master instance and delete the pki directory from each volume mount:
> 
> ```
> sudo rm -rf /mnt/master-vol-*/pki   # mount paths may vary between cloud providers and distros. double check this path
> ```
> 
> 1. Restart the two etcd-manager containers. Alternatively you can reboot the instance or terminate the instance in the autoscaling group
> 
> ```
> sudo docker restart $(sudo docker ps -q -f "label=io.kubernetes.container.name=etcd-manager")
> ```
> 
> When the container restarts, etcd-manager will reissue the certs and rejoin the etcd cluster. You should be able to confirm membership with `etcdctl endpoint status` by following the instructions in the [docs](https://kops.sigs.k8s.io/operations/etcd_administration/#direct-data-access)
> 
> Doing that on one master at a time will avoid any apiserver downtime.
> 
> If anyone has had similar success or is willing to try these instructions please post here. I'm hoping we'll settle on a short term solution at kops office hours tomorrow.

Thank you so much for your sharing @rifelpet , we were lucky to find this page! 
Otherwise we were almost thinking about spinning a new cluster in emergency...

Instructions were clear and worked well, I will just add this extra we needed to do if that could help someone else:
- stop and start the master node instances was needed for us (after the docker restart of etcd-manager containers) otherwise we were still seeing `Unable to create storage backend` & `context deadline exceeded` messages in  /var/log/kube-apiserver.log
- double check the route tables of the master node instances you stopped and started, just to be sure that your correct setup is still there
- if you use **cert-manager** on your cluster, we needed to restart the 3 containers in cert-manager namespace too, in that order : cert-manager-webhook, cert-manager-webhook and cert-manager (before that, we saw a bunch of error messages in those 3 container logs regarding communication with other nodes (like `Failed calling webhook, failing closed certificates.admission.certmanager.k8s.io: failed calling webhook "certificates.admission.certmanager.k8s.io": the server is currently unable to handle the request`) or certificates (like `loading OpenAPI spec for "v1beta1.admission.certmanager.k8s.io" failed with: failed to retrieve openAPI spec, http error: ResponseCode: 503, Body: Error: 'x509: certificate has expired or is not yet valid'`)

Hope it helps!

**Comment by tsturzl** [comment]
Yesterday in an attempt to reissue certificates I ran into some new problems that I don't think have been outlined here. This morning @hakman was assisting me in troubleshooting the issue. What seems to have happened for me is that the secrets that are created for service account tokens were no longer valid after reissuing certs. This caused several services to fail: kube-proxy, dns controller, calico-node and controller, and consequentially caused a cascade of failures for protokube and kubelet of every node. This meant that no nodes would join the cluster, and I had a lot of vital pods refusing to start. While initially it was pretty difficult to narrow down the problem, it seems that the problem was entirely that service account tokens were simply no longer valid and needed reissuing, and this fix is relatively simple. All I had to do was delete all of the secrets of type `kubernetes.io/service-account-token`, and the service account would almost immediately reissue. Once the tokens are reissued you simply have to recreate the pods, in my case by simply deleting them and letting them be recreated, to take up the newly generated tokens. It is my belief that this is a bug, seeing as it took the entire cluster down, and seemingly you would think that in rotating whatever cert is correlated to service accounts that the service account tokens would automatically be reissued.

Additional information:

 - kops version: 1.18.2
 - k8s version: 1.14.0
 - slack thread(kops-users channel): https://kubernetes.slack.com/archives/C3QUFP0QM/p1611160596156500

**Comment by WoodProgrammer** [comment]
Hi there is only one instance is returning this error , I've tried this solution at this link 
https://github.com/kubernetes/kops/issues/8959#issuecomment-618629976 

into only one master-node with the name MASTER-A (cluster has 3 master-node) so etcd was not coming back up and  `tls certificate error still returning reffering to "serverName: other-master-address" `

```
etcdmain: rejected connection from "MASTER-B-ADDRESS:64762" (error "tls: failed to verify client's certificate: x509: certificate has expired or is not yet valid
etcdmain: rejected connection from "MASTER-C-ADDRESS:64762" (error "tls: failed to verify client's certificate: x509: certificate has expired or is not yet valid
```
Do you have any idea about that ? 

Thanks

**Comment by lboix** [comment]
> Hi there is only one instance is returning this error , I've tried this solution at this link
> [#8959 (comment)](https://github.com/kubernetes/kops/issues/8959#issuecomment-618629976)
> 
> into only one master-node with the name MASTER-A (cluster has 3 master-node) so etcd was not coming back up and `tls certificate error still returning reffering to "serverName: other-master-address" `
> 
> ```
> etcdmain: rejected connection from "MASTER-B-ADDRESS:64762" (error "tls: failed to verify client's certificate: x509: certificate has expired or is not yet valid
> etcdmain: rejected connection from "MASTER-C-ADDRESS:64762" (error "tls: failed to verify client's certificate: x509: certificate has expired or is not yet valid
> ```
> 
> Do you have any idea about that ?
> 
> Thanks

Hello @WoodProgrammer, I would suggest to apply the main solution you linked to all of your 3 master nodes.
They have been likely created around the same time (at the cluster creation) so if one master already had its etcd certificates expired because of what we discuss in this thread (etcd-manager version too old that does rotate certificates automatically) then the two others will follow soon.

It's better to start over with fresh new etcd-manager containers on all master nodes.

Hope it helps!

**Comment by yaser-hameed** [comment]
> The certs and keys in S3 are only for the CA certificates which have a 10 year expiration. The individual certs with 1 year expirations are issued locally on each host, so nothing needs to be removed from S3 at the moment.
> 
> We just successfully rotated the etcd certs on our clusters and found these steps to work:
> 
> 1. SSH into each master instance and delete the pki directory from each volume mount:
> 
> ```
> sudo rm -rf /mnt/master-vol-*/pki   # mount paths may vary between cloud providers and distros. double check this path
> ```
> 
>  1. Restart the two etcd-manager containers. Alternatively you can reboot the instance or terminate the instance in the autoscaling group
> 
> ```
> sudo docker restart $(sudo docker ps -q -f "label=io.kubernetes.container.name=etcd-manager")
> ```
> 
> When the container restarts, etcd-manager will reissue the certs and rejoin the etcd cluster. You should be able to confirm membership with `etcdctl endpoint status` by following the instructions in the [docs](https://kops.sigs.k8s.io/operations/etcd_administration/#direct-data-access)
> 
> Doing that on one master at a time will avoid any apiserver downtime.
> 
> If anyone has had similar success or is willing to try these instructions please post here. I'm hoping we'll settle on a short term solution at kops office hours tomorrow.

This worked for me as well.

**Comment by bhaumikmithun** [comment]
> > The certs and keys in S3 are only for the CA certificates which have a 10 year expiration. The individual certs with 1 year expirations are issued locally on each host, so nothing needs to be removed from S3 at the moment.
> > We just successfully rotated the etcd certs on our clusters and found these steps to work:
> > 
> > 1. SSH into each master instance and delete the pki directory from each volume mount:
> > 
> > ```
> > sudo rm -rf /mnt/master-vol-*/pki   # mount paths may vary between cloud providers and distros. double check this path
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 1. Restart the two etcd-manager containers. Alternatively you can reboot the instance or terminate the instance in the autoscaling group
> > 
> > ```
> > sudo docker restart $(sudo docker ps -q -f "label=io.kubernetes.container.name=etcd-manager")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When the container restarts, etcd-manager will reissue the certs and rejoin the etcd cluster. You should be able to confirm membership with `etcdctl endpoint status` by following the instructions in the [docs](https://kops.sigs.k8s.io/operations/etcd_administration/#direct-data-access)
> > Doing that on one master at a time will avoid any apiserver downtime.
> > If anyone has had similar success or is willing to try these instructions please post here. I'm hoping we'll settle on a short term solution at kops office hours tomorrow.
> 
> Thank you so much for your sharing @rifelpet , we were lucky to find this page! Otherwise we were almost thinking about spinning a new cluster in emergency...
> 
> Instructions were clear and worked well, I will just add this extra we needed to do if that could help someone else:
> 
>     * stop and start the master node instances was needed for us (after the docker restart of etcd-manager containers) otherwise we were still seeing `Unable to create storage backend` & `context deadline exceeded` messages in  /var/log/kube-apiserver.log
> 
>     * double check the route tables of the master node instances you stopped and started, just to be sure that your correct setup is still there
> 
>     * if you use **cert-manager** on your cluster, we needed to restart the 3 containers in cert-manager namespace too, in that order : cert-manager-webhook, cert-manager-webhook and cert-manager (before that, we saw a bunch of error messages in those 3 container logs regarding communication with other nodes (like `Failed calling webhook, failing closed certificates.admission.certmanager.k8s.io: failed calling webhook "certificates.admission.certmanager.k8s.io": the server is currently unable to handle the request`) or certificates (like `loading OpenAPI spec for "v1beta1.admission.certmanager.k8s.io" failed with: failed to retrieve openAPI spec, http error: ResponseCode: 503, Body: Error: 'x509: certificate has expired or is not yet valid'`)
> 
> 
> Hope it helps!

i am facing the same issue ... Upgrading to 3.0.20200531  doescnot seems to help
kops version: 1.15.1
k8s version:  1.15.10

Please help here

**Comment by olemarkus** [comment]
Strongly recommend you upgrade to kops 1.21 (last version to support k8s 1.15) to resolve this. Remove any etcd-manager version from the spec.

**Comment by lboix** [comment]
@bhaumikmithun can you connect to the master nodes ? If yes then the two **sudo** commands in your quoted message can help (do one master at the time)

**Comment by bhaumikmithun** [comment]
> @bhaumikmithun can you connect to the master nodes ? If yes then the two **sudo** commands in your quoted message can help (do one master at the time)

@lboix did that but still not able to recover from error

**Comment by olemarkus** [comment]
Did you rotate all the control plane nodes? The etcd-manager upgrade only happens as part of a rolling update.

**Comment by tuapuikia** [comment]
Hi,

is it possible to upgrade kops 1.17 to 1.21 on existing cluster?

**Comment by olemarkus** [comment]
> is it possible to upgrade kops 1.17 to 1.21 on existing cluster?

Yes. As long as your K8s version is supported by kOps 1.21

**Comment by bomee** [comment]
> 

I had a similar problem. But I fixed the etcd certificate issue, the apiserver certificate is expired, can you tell me how to manually generate a certificate using openssl

**Comment by bomee** [comment]
> > > > I would suggest to try to upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531` by editing the manifest files and restarting the etcd pods:
> > > > ```
> > > > /etc/kubernetes/manifests/etcd-events.manifest
> > > > /etc/kubernetes/manifests/etcd-main.manifest
> > > > ```
> > > 
> > > 
> > > @hakman : I also experience similar issue like @duncanhall where I couldn't communicate using kubectl because kupe-apiserver certificate expired. So I manually (using openssl) created the certificate and replaced `/etc/kubernetes/pki/kube-apiserver/etcd-client.crt` and `/etc/kubernetes/pki/kube-apiserver/etcd-client.key` after that kube-apiserver docker started working and I am able to connect using kubectl.
> > > Then I followed the workaround from [https://kops.sigs.k8s.io/advisories/etcd-manager-certificate-expiration/](url) to generate certs for etcd which are stored at /mnt/master-vol-_/pki. Now I have renewed certs at /mnt/master-vol-_/pki but both etcd-manager-main and etcd-manager-events are giving error saying `not able to connect to peer etcd-a` (for main) and `etcd-events-a` (for events). More details on errors are posted on my comments on this thread two days ago.
> > > @hakman: I have not upgraded etcd-manager to `kopeio/etcd-manager:3.0.20200531` yet. I am currently on `kopeio/etcd-manager:3.0.20190516`. As of now, I don't see /var folder under /mnt/master-vol-*/. I have only `data`, `pki`, `state` and `lost+found` folders.
> > > Is it ok for me to upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531`?
> > > @duncanhall : You mentioned that you restored data from a stale backup, could you please provide what folders data you had to restore?
> > > Thanks @hakman @duncanhall
> > 
> > 
> > **Update to my previous comment**
> > I have two other Kubernertes clusters setup using kops in dev/qa. These clusters are running Kubernetes version v1.11.1 and do not have etcd-manager. I checked the path on the master volumes and found the path to be
> > for `etcd.manifest` --> `/mnt/master-vol-***/var/etcd/data/member/` which has folders `snap` and `wal`
> > for `etcd-events.manifest` --> `/mnt/master-vol-****/var/etcd/data-events/member/` which has folders `snap` and `wal`
> > So in my production cluster which has etcd-manager installed (`kopeio/etcd-manager:3.0.20190516`), folders under /mnt/master-vol/*** are `data`, `pki`, `state` and `lost+found` folders.
> > `/mnt/master-vol-****/data/<token>/member/` which has folders `snap` and `wal`
> > Just want to make sure below steps are correct to fix the issue?
> > 
> > * Backup existing `data` and `data-events` folders (or all the contents under /mnt/master-vol-*)
> > * Upgrade etcd-manager to `kopeio/etcd-manager:3.0.20200531` and check if it starts correctly and resolves the issue.
> > * If it complains about missing the folder ``/mnt/master-vol-_**/var/etcd/data`or`/mnt/master-vol-**_/var/etcd/data-events`. then copy the contents from the backup.
> > 
> > Appreciate if anyone could please comment on the above or if I am missing any steps.
> > Many thanks.
> 
> **Update:**
> 
> I have upgraded etcd-manager to `kopeio/etcd-manager:3.0.20200531` which resolved the issue. Cluster is back to healthy state. Many thanks to @hakman.

I had a similar problem. But I fixed the etcd certificate issue, the apiserver certificate is expired, can you tell me how to manually generate a certificate using openssl
